sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/collectionOperations.scala

```diff
@@ -4600,3 +4600,69 @@ case class ArrayExcept(left: Expression, right: Expression) extends ArrayBinaryL
   override protected def withNewChildrenInternal(
     newLeft: Expression, newRight: Expression): ArrayExcept = copy(left = newLeft, right = newRight)
 }
+
+/**
+ * Given an array, and another element append the element at the end of the array.
+ */
+@ExpressionDescription(
+  usage = "_FUNC_(expr, expr) - Prepend the element",
+  examples = """
+    Examples:
+      > SELECT _FUNC_(array('b', 'd', 'c', 'a'), array(1, 2, 3, 4));
+
+  """,
+  since = "3.4.0",
+  group = "collection_funcs")
+case class ArrayAppend(left: Expression, right: Expression)
+    extends BinaryExpression
+    with ImplicitCastInputTypes {
+  override def prettyName: String = "array_append"
+  override def inputTypes: Seq[AbstractDataType] = Seq(ArrayType, AnyDataType)
+
+  override def checkInputDataTypes(): TypeCheckResult = {
+    (left.dataType, right.dataType) match {
+      case (ArrayType(e1, _), (e2)) if e1.sameType(e2) =>
+        TypeCheckResult.TypeCheckSuccess
+      case _ =>
```

Comment:
https://docs.snowflake.com/en/sql-reference/functions/array_append.html

Refer to the description document of snowflake, The new element data type does not need to match the data type(s) of the existing elements in the array.

So `ARRAY_ELEMENT_DIFF_TYPES` should not be returned

Reply:
@LuciferYang Thanks for the review! Arrays are more strongly typed in Spark ! In case when we are getting mixed types of elements to append in the array, should we change the type of the array itself ?
Also, we cannot create a column with Array[AnyDataType] like below, we will eventually get an error

`scala> val df3 = Seq((Array("a", "b", 2, 5d), 3)).toDF("a", "b")
org.apache.spark.SparkUnsupportedOperationException: No Encoder found for Any
- array element class: "java.lang.Object"
- field (class: "scala.Array", name: "_1")
- root class: "scala.Tuple2"
`

Which is why I thought that its better to analyze the types of the array at the start itself and raise if the types do not match.

Reply:
This is snowflake example:
```
select array_append(array_construct(1, 2, 3), 'HELLO');
+-------------------------------------------------+
| ARRAY_APPEND(ARRAY_CONSTRUCT(1, 2, 3), 'HELLO') |
|-------------------------------------------------|
| [                                               |
|   1,                                            |
|   2,                                            |
|   3,                                            |
|   "HELLO"                                       |
| ]                                               |
+-------------------------------------------------+
```

I think we should keep up with it? WDYT @zhengruifeng @HyukjinKwon

Reply:
I think snowflake's arrays are not strongly typed while spark's arrays' are strongly typed, since it does not allow us to mix multiple types of elements in one Array. But yes if we want to make it similar to snowflake, we will have to make spark arrays also loosely typed. Let's wait for everyone else's suggestions as well. And I think it will be a similar construct for array_insert, array_prepend as well. :)

Reply:
Documenting the behaviour of snowflake for this function in different scenarios
```
select array_append(array_construct(1, 2, 3), 'HELLO');
-- [ 1, 2, 3, "HELLO" ]


select array_append(array_construct(1, 2, 3), NULL);
-- [1,2,3,undefined]


select array_append(NULL, 'a');
-- null

select array_append(NULL, NULL);
-- null 
```
So, if the array is null, we don't create an array and return null directly while if the element is null we just insert it directly without checking if it is null.

As of now I think the function is Null Intolerant, with any null value it will return null. Need to override this behaviour.

Reply:
For the case 

```
select array_append(array_construct(1, 2, 3), 'HELLO'); -- [ 1, 2, 3, "HELLO" ]
```

Seems that Spark cannot give the same semantics, WDYT @HyukjinKwon

Reply:
For this case, I think we can refer to `ArrayUnion`:

```
scala> spark.sql(""" SELECT array_union(a,b) FROM VALUES (ARRAY(1,2,3), ARRAY(8,9), ARRAY('HELLO')) AS tab(a,b,c) """)
res7: org.apache.spark.sql.DataFrame = [array_union(a, b): array<int>]

scala> spark.sql(""" SELECT array_union(a,c) FROM VALUES (ARRAY(1,2,3), ARRAY(8,9), ARRAY('HELLO')) AS tab(a,b,c) """)
org.apache.spark.sql.AnalysisException: [DATATYPE_MISMATCH.BINARY_ARRAY_DIFF_TYPES] Cannot resolve "array_union(a, c)" due to data type mismatch: Input to function `array_union` should have been two "ARRAY" with same element type, but it's ["ARRAY<INT>", "ARRAY<STRING>"].; line 1 pos 8;
'Project [unresolvedalias(array_union(a#13, c#15), None)]
+- SubqueryAlias tab
   +- LocalRelation [a#13, b#14, c#15]

```

I think we should apply same datatype validation as `ArrayUnion`

Reply:
also cc @HyukjinKwon @cloud-fan

Reply:
We have no other choice. Spark array type only allows one element type. We can find the wider type the array element type and the data type of the to-be-added value.